src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/playback/GerritMissedEventsPlaybackManager.java

```diff
@@ -0,0 +1,486 @@
+
+        try {
+            load();
+        } catch (IOException e) {
+            logger.error("Failed to load in timestamps for server {}", serverName);
+            logger.error("Exception: {}", e.getMessage(), e);
+            playBackComplete = true;
+            return;
+        }
+        Date timeStampDate = getDateFromTimestamp();
+        long diff = new Date().getTime() - timeStampDate.getTime();
+        if (diff > 0) {
+            logger.debug("Non-zero date range from last-alive timestamp exists for server " + serverName
+                    + ": " + Util.getPastTimeString(diff));
+        } else {
+            logger.debug("Zero date range from last-alive timestamp for server " + serverName);
+            playBackComplete = true;
+            return;
+        }
+        try {
+            List<GerritTriggeredEvent> events = getEventsFromDateRange(timeStampDate);
+            logger.info("({}) missed events to process for server: {} ...", events.size(), serverName);
+            for (GerritTriggeredEvent evt: events) {
+                logger.info("({}) Processing missed event {}", serverName, evt);
+                boolean receivedEvtFound = false;
+                synchronized (receivedEventCache) {
+                  Iterator<GerritTriggeredEvent> i = receivedEventCache.iterator(); // Must be in synchronized block
+                  while (i.hasNext()) {
+                      GerritTriggeredEvent rEvt = i.next();
+                      if (rEvt.equals(evt)) {
+                        receivedEvtFound = true;
+                        break;
+                      }
+                  }
+                }
+                if (receivedEvtFound) {
+                    logger.debug("({}) Event already triggered...skipping trigger.", serverName);
+                } else {
+
+                    //do we have this event in the time slice?
+                    long currentEventCreatedTime = ((GerritTriggeredEvent)evt).getEventCreatedOn().getTime();
+                    if (serverTimestamp.getTimeSlice() == currentEventCreatedTime) {
+                        if (serverTimestamp.getEvents().contains(evt)) {
+                            logger.debug("({}) Event already triggered from time slice...skipping trigger.", serverName);
+                            continue;
+                        }
+                    }
+                    logger.info("({}) Triggering: {}", serverName, evt);
+                    GerritServer server = PluginImpl.getServer_(serverName);
+                    if (server == null) {
+                        logger.error("Server for {} could not be found. Skipping this event", serverName);
+                        continue;
+                    }
+                    server.triggerEvent(evt);
+                    receivedEventCache.add(evt);
+                    logger.debug("Added event {} to received cache for server: {}", evt, serverName);
+                    persist(evt);
```

Comment:
We need some way of knowing what timestamp to use to request missed events from Gerrit when the jenkins instance comes back up. At some point, we need to persist the timestamp.

I can look into something like MapDB that maybe more performant.

Do you have any ideas? I'll gladly check them out.